### PR TITLE
Staking withdraw fixes

### PR DESCRIPTION
**Pull request by jameshowlett977**
Fixes for the following issues:
1. It wasn't possible to withdraw stake/extend stake/delegate voting power:
delegate checkpoint wasn't updating for the second and next stakes for the same date if the first stake was withdrawn completely and the second stake was delegating to the staker (no delegation to another address).
2. Vesting contract can't withdraw stakes if it has stakes between locking dates for the first stake: Vesting contract iterates stakes using 4 weeks.
3. Multisig should be able to invoke governanceWithdrawVesting. It impacts only team vesting contracts.

### CODE REVIEW

**Comment by jjmr007** [comment]
User reports related to extending stakes and withdraw fron Stakes contract:

--------------------------------

Freshdesk [Ticket #105](https://sovryn.freshdesk.com/a/tickets/105):

Fjodor reported via the portal
25 days ago (Sat, 8 May 2021 at 8:10 AM)
 
I'm trying to extend staking but those transactions fail on the rsk explorer.
Then I tried withdrawing, but those transactions also fail.
I tried different computers and different browsers (Chrome, Edge and Brave).
Even made a new virtual machine to install clean browsers and Liquality plugin. I don't know what to do next.

For example:
https://explorer.rsk.co/tx/0xfa4ca495de3ba605d415e01b4d297d4cdd2bead1599ba7f17f28c82de6cb9180
https://explorer.rsk.co/tx/0x95bd2cbedf755fec124e9dadca8292e5330015e573f18f6898c1ae22c8daada3
https://explorer.rsk.co/tx/0xd146573f000d174f9ce8e5dd72d06dcee3427846c58a0a4e7108015d35ee74ce
https://explorer.rsk.co/tx/0x2829efa350689d81c6c695ba5feb4accc606ae442da1209140014d7d1ce4a5d3
https://explorer.rsk.co/tx/0x273eae366286804d155da192ec6bc6dcfc26711d3627fb36c835248804fa9ab0
I'm using Liquality 0.29.0.

-------------------------------------------------------------------------------------------------------


Freshdesk [Ticket #239](https://sovryn.freshdesk.com/a/tickets/239):

luis miguel reported via the portal
15 days ago (Tue, 18 May 2021 at 4:01 PM)
 
I can't unstake the 100 SOVS or extending the contract, always saying failed transaction.
txid https://explorer.rsk.co/tx/0x5a4bc108b18c8d998ff2f2c5b49b7a39a829965de2d5356d705bb3cb22cb1e57
my rsk wallet
0x3E07df76bCDa0f20514cAE0350E123812438A998
Captura de pantalla 2021-05-18 a las 17 49 55

-------------------------------------------------------------------------------------------

Freshdesk [Ticket #244](https://sovryn.freshdesk.com/a/tickets/244):

Ulises Guareschi Corvetto reported via the portal
15 days ago (Tue, 18 May 2021 at 9:20 PM)
 
You can see my address (0x3BF76665B8f364B285B391Bb9fCf17dd10d2F427). All the transactions fails.

My answer to this user:

Hi Ulises Guareschi Corvetto,

Regarding this transaction with the Staking contract:

https://explorer.rsk.co/tx/0x9efc1a000afd0598e0fbfd692d657ed8f1080215c0d0a369193b1820aa8dfc33

There is a very serious bug we are tying to solve ASAP. ETA is around 2 weeks. Your funds are safe, but stuk for now.

Regarding transactions like these: 
​
https://explorer.rsk.co/tx/0x1fa15f2d5f8b47458e19debea02cee11ecce78107d760b2fbb06183bdf2d2481
https://explorer.rsk.co/tx/0xe282fbdb4b4b43bd8014cb1cd8479b0360c98b824f43cbd691c5fabf4b5f1cd2
and so on...

All has the same error:

Error Message:
The provided amount should be identical to msg.value
​
​It is most likely a UI bug which I've just reported: https://github.com/DistributedCollective/Sovryn-frontend/issues/1125

Sorry for the inconveniences

Regards
J.Moros

-------------------------------------------------------------------------

That's all so far... If I found more reports I'll edit this comment. Thanks you all.

**Comment by jjmr007** [comment]
User reports related to extending stakes and withdraw fron Stakes contract:

--------------------------------

Ticket N°105:

I'm trying to extend staking but those transactions fail on the rsk explorer.
Then I tried withdrawing, but those transactions also fail.
I tried different computers and different browsers (Chrome, Edge and Brave).
Even made a new virtual machine to install clean browsers and Liquality plugin. I don't know what to do next.

For example:
https://explorer.rsk.co/tx/0xfa4ca495de3ba605d415e01b4d297d4cdd2bead1599ba7f17f28c82de6cb9180
https://explorer.rsk.co/tx/0x95bd2cbedf755fec124e9dadca8292e5330015e573f18f6898c1ae22c8daada3
https://explorer.rsk.co/tx/0xd146573f000d174f9ce8e5dd72d06dcee3427846c58a0a4e7108015d35ee74ce
https://explorer.rsk.co/tx/0x2829efa350689d81c6c695ba5feb4accc606ae442da1209140014d7d1ce4a5d3
https://explorer.rsk.co/tx/0x273eae366286804d155da192ec6bc6dcfc26711d3627fb36c835248804fa9ab0
I'm using Liquality 0.29.0.

-------------------------------------------------------------------------------------------------------


Ticket N°239


I can't unstake the 100 SOVS or extending the contract, always saying failed transaction.
txid https://explorer.rsk.co/tx/0x5a4bc108b18c8d998ff2f2c5b49b7a39a829965de2d5356d705bb3cb22cb1e57
my rsk wallet
0x3E07df76bCDa0f20514cAE0350E123812438A998

-------------------------------------------------------------------------------------------

Ticket N°244:


You can see my address (0x3BF76665B8f364B285B391Bb9fCf17dd10d2F427). All the transactions fails.

My answer to this user:

Regarding this transaction with the Staking contract:

https://explorer.rsk.co/tx/0x9efc1a000afd0598e0fbfd692d657ed8f1080215c0d0a369193b1820aa8dfc33

There is a very serious bug we are tying to solve ASAP. ETA is around 2 weeks. Your funds are safe, but stuk for now.

Regarding transactions like these: 

https://explorer.rsk.co/tx/0x1fa15f2d5f8b47458e19debea02cee11ecce78107d760b2fbb06183bdf2d2481
https://explorer.rsk.co/tx/0xe282fbdb4b4b43bd8014cb1cd8479b0360c98b824f43cbd691c5fabf4b5f1cd2
and so on...

All has the same error:

Error Message:
The provided amount should be identical to msg.value

It is most likely a UI bug which I've just reported: https://github.com/DistributedCollective/Sovryn-frontend/issues/1125

Sorry for the inconveniences

Regards
J.Moros

-------------------------------------------------------------------------

That's all so far... If I found more reports I'll edit this comment. Thanks you all.

**Comment by jjmr007** [comment]
User reports related to extending stakes and withdraw fron Stakes contract:

--------------------------------

Ticket N°105:

I'm trying to extend staking but those transactions fail on the rsk explorer.
Then I tried withdrawing, but those transactions also fail.
I tried different computers and different browsers (Chrome, Edge and Brave).
Even made a new virtual machine to install clean browsers and Liquality plugin. I don't know what to do next.

For example:
https://explorer.rsk.co/tx/0xfa4ca495de3ba605d415e01b4d297d4cdd2bead1599ba7f17f28c82de6cb9180
https://explorer.rsk.co/tx/0x95bd2cbedf755fec124e9dadca8292e5330015e573f18f6898c1ae22c8daada3
https://explorer.rsk.co/tx/0xd146573f000d174f9ce8e5dd72d06dcee3427846c58a0a4e7108015d35ee74ce
https://explorer.rsk.co/tx/0x2829efa350689d81c6c695ba5feb4accc606ae442da1209140014d7d1ce4a5d3
https://explorer.rsk.co/tx/0x273eae366286804d155da192ec6bc6dcfc26711d3627fb36c835248804fa9ab0
I'm using Liquality 0.29.0.

-------------------------------------------------------------------------------------------------------


Ticket N°239


I can't unstake the 100 SOVS or extending the contract, always saying failed transaction.
txid https://explorer.rsk.co/tx/0x5a4bc108b18c8d998ff2f2c5b49b7a39a829965de2d5356d705bb3cb22cb1e57
my rsk wallet
0x3E07df76bCDa0f20514cAE0350E123812438A998

-------------------------------------------------------------------------------------------

Ticket N°244:


You can see my address (0x3BF76665B8f364B285B391Bb9fCf17dd10d2F427). All the transactions fails.

My answer to this user:

Regarding this transaction with the Staking contract:

https://explorer.rsk.co/tx/0x9efc1a000afd0598e0fbfd692d657ed8f1080215c0d0a369193b1820aa8dfc33

There is a very serious bug we are tying to solve ASAP. ETA is around 2 weeks. Your funds are safe, but stuk for now.

Regarding transactions like these: 

https://explorer.rsk.co/tx/0x1fa15f2d5f8b47458e19debea02cee11ecce78107d760b2fbb06183bdf2d2481
https://explorer.rsk.co/tx/0xe282fbdb4b4b43bd8014cb1cd8479b0360c98b824f43cbd691c5fabf4b5f1cd2
and so on...

All has the same error:

Error Message:
The provided amount should be identical to msg.value

It is most likely a UI bug which I've just reported: https://github.com/DistributedCollective/Sovryn-frontend/issues/1125

Sorry for the inconveniences

Regards


-------------------------------------------------------------------------

That's all so far... If I found more reports I'll edit this comment. Thanks you all.

File: contracts/governance/Staking/Staking.sol [hunk 1]
```diff
@@ -323,6 +366,24 @@ contract Staking is IStaking, WeightedStaking, ApprovalReceiver {
...
+		uint96 amount,
+		uint256 until,
+		address receiver,
+		bool isGovernance
+	) internal {
+		if (msg.sender.isContract()) {
+			uint256 nextLock = until.add(TWO_WEEKS);
+			if (isGovernance || block.timestamp >= nextLock) {
+				uint96 stake = _getPriorUserStakeByDate(msg.sender, nextLock, block.number - 1);
+				if (stake > 1) {
```
Comment (eMarchenko) [inline contracts/governance/Staking/Staking.sol:380]: same here, `_getPriorUserStakeByDate`  returns 0 correctly, so the if-clause should be `if (stake > 0) {`

File: contracts/governance/Staking/Staking.sol [hunk 2]
```diff
@@ -323,6 +366,24 @@ contract Staking is IStaking, WeightedStaking, ApprovalReceiver {
...
 	}
 
+	// @dev withdraws tokens for lock date 2 weeks later than given lock date
+	function _withdrawNext(
+		uint96 amount,
+		uint256 until,
+		address receiver,
+		bool isGovernance
+	) internal {
+		if (msg.sender.isContract()) {
```
Comment (eMarchenko) [inline contracts/governance/Staking/Staking.sol:376]: is it an attempt to identify vesting contract? please be more specific

File: contracts/governance/Staking/Staking.sol [hunk 3]
```diff
@@ -490,6 +557,28 @@ contract Staking is IStaking, WeightedStaking, ApprovalReceiver {
...
 	}
 
+	// @dev delegates tokens for lock date 2 weeks later than given lock date
+	//		if message sender is a contract
+	function _delegateNext(
+		address delegator,
+		address delegatee,
+		uint256 lockedTS
+	) internal {
+		if (msg.sender.isContract()) {
```
Comment (eMarchenko) [inline contracts/governance/Staking/Staking.sol:567]: anyone can deploy a contract to get access to that logic. perform a more specific check

File: contracts/governance/Staking/Staking.sol [hunk 4]
```diff
@@ -490,6 +557,28 @@ contract Staking is IStaking, WeightedStaking, ApprovalReceiver {
...
+			address currentDelegate = delegates[delegator][nextLock];
+			if (currentDelegate != delegatee) {
+				_delegate(delegator, delegatee, nextLock);
+			}
+			// @dev workaround for the issue with a delegation of the latest stake
+			nextLock = nextLock.add(TWO_WEEKS);
+			currentDelegate = delegates[delegator][nextLock];
+			if (currentDelegate != delegatee) {
+				_delegate(delegator, delegatee, nextLock);
+			}
```
Comment (eMarchenko) [inline contracts/governance/Staking/Staking.sol:578]: this block duplicates lines above it. is it correct? basically, it's a loop with 2 iterations, where `2` is a magic constant - doesn't feel right.

**Review by eMarchenko — COMMENTED** [review]
Consider saving vesting address and only triggering hack logic when called by it

File: contracts/governance/Staking/Staking.sol [hunk 5]
```diff
@@ -290,6 +328,11 @@ contract Staking is IStaking, WeightedStaking, ApprovalReceiver {
 		address receiver,
 		bool isGovernance
 	) internal {
+		// @dev it's very unlikely some one will have 1/10**18 SOV staked in Vesting contract
+		//		this check is a part of workaround for Vesting.withdrawTokens issue
+		if (amount == 1 && msg.sender.isContract()) {
```
[the remark this answers is not in the document]
Comment (jameshowlett977) [reply contracts/governance/Staking/Staking.sol:333]: We have an additional invocation stack:

VestingLogic(vesting address).withdrawTokens
stake = staking.getPriorUserStakeByDate(address(this), i, block.number - 1);

getPriorUserStakeByDate will return 1
_withdraw(amount = 1 ...)

[the remark this answers is not in the document]
Comment (eMarchenko) [reply contracts/governance/Staking/Staking.sol:333]: oh I see

File: contracts/governance/Staking/Staking.sol (hunk 1 above)
Comment (jameshowlett977) [reply contracts/governance/Staking/Staking.sol]: You're right, it doesn't make sense here,
will be checkde in _withdraw anyway.
Changed to 0.

File: contracts/governance/Staking/WeightedStaking.sol [hunk 6]
```diff
@@ -266,7 +268,28 @@ contract WeightedStaking is Checkpoints {
 		address account,
 		uint256 date,
 		uint256 blockNumber
-	) public view returns (uint96) {
+	) external view returns (uint96) {
+		uint96 priorStake = _getPriorUserStakeByDate(account, date, blockNumber);
+		// @dev we need to modify function in order to workaround issue with Vesting.withdrawTokens:
+		//		return 1 instead of 0 if message sender is a contract.
+		if (priorStake == 0 && msg.sender.isContract()) {
```
[the remark this answers is not in the document]
Comment (jameshowlett977) [reply contracts/governance/Staking/WeightedStaking.sol:275]: added function _isVestingContract

File: contracts/governance/Staking/Staking.sol (hunk 2 above)
Comment (jameshowlett977) [reply contracts/governance/Staking/Staking.sol]: added function _isVestingContract

File: contracts/governance/Staking/Staking.sol (hunk 3 above)
Comment (jameshowlett977) [reply contracts/governance/Staking/Staking.sol]: added function _isVestingContract

File: contracts/governance/Staking/Staking.sol (hunk 4 above)
Comment (jameshowlett977) [reply contracts/governance/Staking/Staking.sol]: Added test to reproduce an issue with delegation,
it fails without this delegation.
Also added additional check before function invocation.

**Comment by jameshowlett977** [comment]
@eMarchenko Thank you for the review.
I've applied part of the comments and explained why we can't fix others.
Please take a look.

**Review by eMarchenko — APPROVED** [review]

File: contracts/governance/Staking/WeightedStaking.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (ororopickpocket) [reply contracts/governance/Staking/WeightedStaking.sol:275]: but here, you're still calling msg.sender.isContract(). Why not isVestingContract?

[the remark this answers is not in the document]
Comment (jameshowlett977) [reply contracts/governance/Staking/WeightedStaking.sol]: oh, thank you, fixed

**Comment by jjmr007** [comment]
I found a couple of cases of Origin-Vesting-Logics with unlock date of 1622802495 (june-4) which are unable to withdraw tokens:

"dates1":["1622802495"],
"stakes1":["68895000000000000000"],
"user_Origin1":"0x23e9fd8572e5ad4651c0a0a5436d872810dfaf02",
"Vesting_Origin1":"0x3330C1dE5ED7ec70aBB2A1Dd519A53F93B14037A"

"dates2":["1622802495"],
"stakes2":["531810000000000000000"],
"user_Origin2":"0xb92067c703d30b75ae3fef020dd6f62869e26355",
"Vesting_Origin2":"0xB943F2194977ADa97Ef85c9784821b4C38a6d65B"

**Comment by jjmr007** [comment]
This problem persist:

Before: "Vesting contract can't withdraw stakes if it has stakes between locking dates for the first stake: Vesting contract iterates stakes using 4 weeks."

Now: Vesting contract can't withdraw stakes if it has stakes between locking dates for the **second** stake.

Example: for this vesting contract `0x511ca82608F08F680e7cb8Aa909317D1E41f091E` it most imminent unlock date is `1624012095` = `GMT: Friday, 18 June 2021 10:28:15` so the withdrawTokens method should result in a token transfer, but it didn't happen, as can be seen in [this transaction](https://explorer.testnet.rsk.co/tx/0xc3f31c55911f908c812d2ab2247229893dfb4c3f66646734bee454da26fc5f27).

**Pull request merged**